README.md
# Credit Card with Datomic

This project was created for study purpose.

It simulates a database on [Datomic](https://www.datomic.com/).


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2021-12-10
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2021-12-10
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/credit-card-datomic/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/credit-card-datomic/compare/0.1.0...0.1.1


test/credit_card_datomic/core_test.clj
(ns credit-card-datomic.core-test
  (:require [clojure.test :refer :all]
            [credit-card-datomic.core :refer :all]))

(deftest a-test
  (testing "FIXME, I fail."
    (is (= 0 1))))


test/credit_card_datomic/client_test.clj
(ns credit-card-datomic.client-test
  (:require [clojure.test :refer :all]
            [credit-card-datomic.client :refer :all])
  (:import (clojure.lang ExceptionInfo)))

(deftest new-client-test
  (testing "Should add new client successfully"
    (is (= {:client/name "Foo"
            :client/cpf  "400.111.222-30"
            :client/card []}
           (dissoc (new-client "Foo" "400.111.222-30") :client/id)))
    (is (not (nil? (:client/id (new-client "Foo"
                                           "400.111.222-30"))))))
  (testing "Schema not create"
    (is (thrown? ExceptionInfo (new-client 111 "400.111.222-30")))
    (is (thrown? ExceptionInfo (new-client "Foo" 40011122230)))))


test/credit_card_datomic/credit_card_test.clj
(ns credit-card-datomic.credit-card-test
  (:require [clojure.test :refer :all]
            [credit-card-datomic.credit-card :refer :all])
  (:import (clojure.lang ExceptionInfo)))

(deftest new-card-test
  (testing "Schema created successfully"
    (is (= {:card/number          "5555 5555 5555 555"
            :card/cvv             123
            :card/expiration-date "10/29"}
          (dissoc (new-card "5555 5555 5555 555"
                            123
                            "10/29") :card/id)))
    (is (not (nil? (:card/id (new-card "5555 5555 5555 555"
                                       123
                                       "10/29"))))))
  (testing "Schema not create"
    (is (thrown? ExceptionInfo (new-card 5555 123 "10/29")))
    (is (thrown? ExceptionInfo (new-card "5555 5555 5555 555" "123" "10/29")))
    (is (thrown? ExceptionInfo (new-card "5555 5555 5555 555" 123 10)))))


src/credit_card_datomic/db.clj
(ns credit-card-datomic.db
  (:require [datomic.api :as d]))

(def db-uri "datomic:dev://localhost:4334/credit-card")

(defn open-db  []
  (d/create-database db-uri)
  (d/connect db-uri))

(defn delete-db []
  (d/delete-database db-uri))

(def conn (open-db))

(defn create-schema! [schema]
  (d/transact conn schema))

(defn upsert-on-db!
  [data]
  (d/transact conn data))

(defn all-data!
  ([sent-key]
   (all-data! sent-key (d/db conn)))
  ([sent-key db]
   (d/q '[:find [(pull ?entity [*]) ...]
          :in $ ?key
          :where [?entity ?key]] db sent-key)))

(defn one-item!
  ([wanted-key wanted-data]
   (one-item! wanted-key wanted-data (d/db conn)))
  ([wanted-key wanted-data db]
   (d/pull db '[*] [wanted-key wanted-data])))

(defn many-itens!
  ([wanted-key wanted-data]
   (many-itens! wanted-key wanted-data (d/db conn)))
  ([wanted-key wanted-data db]
   (d/q '[:find [(pull ?entity [*]) ...]
          :in $ ?key ?data
          :where [?entity ?key ?data]]
        db wanted-key wanted-data)))

(defn clients-expends!
  ([sent-client-id]
   (clients-expends! sent-client-id (d/db conn)))
  ([sent-client-id db]
   (d/q '[:find [(pull ?expend [*])...]
          :in $ ?client-id
          :where [?client :client/id ?client-id]
          [?client :client/card ?card]
          [?expend :expend/card ?card]]
        db sent-client-id)))


; (delete-db)








src/credit_card_datomic/client.clj
(ns credit-card-datomic.client
  (:require [schema.core :as s]
            [credit-card-datomic.credit-card :as c.credit-card]
            [credit-card-datomic.db :as c.db]))

(s/set-fn-validation! true)

(defn uuid [] (java.util.UUID/randomUUID))

(def Client {:client/id                    s/Uuid
             (s/optional-key :client/name) s/Str
             (s/optional-key :client/cpf)  s/Str
             (s/optional-key :client/test)  [s/Str]
             (s/optional-key :client/card) [c.credit-card/Card]})

(def client-schema [{:db/ident       :client/id
                     :db/valueType   :db.type/uuid
                     :db/cardinality :db.cardinality/one
                     :db/unique      :db.unique/identity}
                    {:db/ident       :client/name
                     :db/valueType   :db.type/string
                     :db/cardinality :db.cardinality/one}
                    {:db/ident       :client/cpf
                     :db/valueType   :db.type/string
                     :db/cardinality :db.cardinality/one}
                    {:db/ident       :client/card
                     :db/valueType   :db.type/ref
                     :db/cardinality :db.cardinality/many}])


(c.db/create-schema! client-schema)

(s/defn new-client :- Client
  [name cpf]
  {:client/id   (uuid)
   :client/name name
   :client/cpf  cpf
   :client/test []
   :client/card []})

(s/defn add-clients!
  [clients :- [Client]]
  (c.db/upsert-on-db! clients))

(def client1 (new-client "Client1" "000000000-00"))
(def client2 (new-client "Client2" "111111111-11"))
(def client3 (new-client "Client3" "222222222-22"))
(def card1 (c.credit-card/new-card "5555 4444 3333 2222" 123 "10/28"))
(c.credit-card/add-credit-cards! [card1])

(add-clients! [client1 client2 client3])

(s/defn add-card-on-client!
  [client card]
  (c.db/upsert-on-db! [[:db/add [:client/id (:client/id client)]
                                :client/card
                                [:card/id (:card/id card)]]]))

(s/defn all-clients! []
  (c.db/all-data! :client/id))






src/credit_card_datomic/credit_card.clj
(ns credit-card-datomic.credit-card
  (:use clojure.pprint)
  (:require [schema.core :as s]
            [credit-card-datomic.db :as c.db]
            [datomic.api :as d]))

(s/set-fn-validation! true)

(defn uuid [] (java.util.UUID/randomUUID))

(def Card {:card/id                               s/Uuid
           (s/optional-key :card/number)          s/Str
           (s/optional-key :card/cvv)             Long
           (s/optional-key :card/expiration-date) s/Str})

(def card-schema [{:db/ident       :card/id
                   :db/valueType   :db.type/uuid
                   :db/cardinality :db.cardinality/one
                   :db/unique      :db.unique/identity}
                  {:db/ident       :card/number
                   :db/valueType   :db.type/string
                   :db/cardinality :db.cardinality/one}
                  {:db/ident       :card/expiration-date
                   :db/valueType   :db.type/string
                   :db/cardinality :db.cardinality/one}
                  {:db/ident       :card/cvv
                   :db/valueType   :db.type/long
                   :db/cardinality :db.cardinality/one}])

(c.db/create-schema! card-schema)

(s/defn new-card :- Card
  [number cvv exp-date]
  {:card/id              (uuid)
   :card/number          number
   :card/cvv             cvv
   :card/expiration-date exp-date})

(s/defn add-credit-cards!
  [cards :- [Card]]
  (c.db/upsert-on-db! cards))

(s/defn all-cards []
  (c.db/all-data! :card/id))






















src/credit_card_datomic/expends.clj
(ns credit-card-datomic.expends
  (:use clojure.pprint)
  (:require [schema.core :as s]
            [credit-card-datomic.db :as c.db]
            [datomic.api :as d]
            [credit-card-datomic.credit-card :as c.credit-card]
            [java-time :as t]))

(s/set-fn-validation! true)

(defn uuid [] (java.util.UUID/randomUUID))

(def Expend {:expend/id                        s/Uuid
             (s/optional-key :expend/category) s/Str
             (s/optional-key :expend/value)    BigDecimal
             (s/optional-key :expend/date)     java.util.Date
             (s/optional-key :expend/card)     c.credit-card/Card})


(def expend-schema [{:db/ident       :expend/id
                     :db/valueType   :db.type/uuid
                     :db/cardinality :db.cardinality/one
                     :db/unique      :db.unique/identity}
                    {:db/ident       :expend/category
                     :db/valueType   :db.type/string
                     :db/cardinality :db.cardinality/one}
                    {:db/ident       :expend/value
                     :db/valueType   :db.type/bigdec
                     :db/cardinality :db.cardinality/one}
                    {:db/ident       :expend/date
                     :db/valueType   :db.type/instant
                     :db/cardinality :db.cardinality/one}
                    {:db/ident       :expend/card
                     :db/valueType   :db.type/ref
                     :db/cardinality :db.cardinality/one}])

(c.db/create-schema! expend-schema)

(s/defn new-expend :- Expend
  [category value]
  {:expend/id       (uuid)
   :expend/category category
   :expend/value    value
   :expend/date     (t/java.util.Date.)})

(def card3 (c.credit-card/new-card "5555 4444 3333 2222" 123 "10/28"))
(c.credit-card/add-credit-cards! [card3])

(def expend1 (new-expend "food" 100.00M))
(def expend2 (new-expend "food" 60.00M))

(s/defn add-expends!
  [expends :- [Expend]]
  (c.db/upsert-on-db! expends))

(s/defn add-card-on-expend!
  [expend card]
  (c.db/upsert-on-db! [[:db/add [:expend/id (:expend/id expend)]
                        :expend/card
                        [:card/id (:card/id card)]]]))

(s/defn all-expends! []
  (c.db/all-data! :expend/id))

